File: js-jest/src/gilded_rose.js
class Item {
  constructor(name, sellIn, quality){
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class Shop {
  constructor(items=[]){
    this.items = items;
  }

  subtractOne(x) {
    return x - 1;
  }

  addOne(x) {
    return x + 1;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
        if (this.items[i].name != 'Conjured') {
          if (this.items[i].name != 'Aged Brie' && this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert' && this.items[i].quality > 0) {
            this.items[i].quality = this.subtractOne(this.items[i].quality);
          } else {
            if (this.items[i].quality < 50) {
              this.items[i].quality = this.addOne(this.items[i].quality);
  
              if (this.items[i].name == 'Backstage passes to a TAFKAL80ETC concert') {
                if (this.items[i].sellIn < 11 && this.items[i].quality < 50) {
                  this.items[i].quality = this.addOne(this.items[i].quality);
                }
  
                if (this.items[i].sellIn < 6 && this.items[i].quality < 50) {
                  this.items[i].quality = this.addOne(this.items[i].quality);
                }
              }
            }
          }
        } else {
          this.items[i].quality = this.items[i].quality - 2;
        }

        this.items[i].sellIn = this.items[i].sellIn - 1;

        if (this.items[i].sellIn < 0) {
          if (this.items[i].name != 'Aged Brie') {
            if (this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
              if (this.items[i].quality > 0) {
                this.items[i].quality = this.subtractOne(this.items[i].quality);
              }
            } else {
              this.items[i].quality = this.items[i].quality - this.items[i].quality;
            }
          } else {
            if (this.items[i].quality < 50) {
              this.items[i].quality = this.addOne(this.items[i].quality);
            }
          }
        }
      }
      }

    return this.items;
  }
}

module.exports = {
  Item,
  Shop
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

items = ["Aged Brie", "Backstage passes to a TAFKAL80ETC concert"]
class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if item.name != "Sulfuras, Hand of Ragnaros":
                item.sell_in = item.sell_in - 1
                if item.quality < 50:
                    if item.name in items:
                        if item.sell_in > 0:
                            if item.name == "Aged Brie":
                                item.quality = item.quality + 1
                            elif item.name == "Backstage passes to a TAFKAL80ETC concert":
                                if item.sell_in < 6:
                                    item.quality = item.quality + 3
                                elif item.sell_in < 10:
                                    item.quality = item.quality + 2
                                else:
                                    item.quality = item.quality + 1
                        else:
                            if item.name == "Aged Brie":
                                item.quality = item.quality + 2
                            elif item.name == "Backstage passes to a TAFKAL80ETC concert":
                                item.quality = item.quality - item.quality
                    elif item.quality > 0:
                        if item.sell_in < 0:
                            item.quality = item.quality - 2
                        else:
                            item.quality = item.quality - 1
                elif item.name == "Backstage passes to a TAFKAL80ETC concert" and item.sell_in < 1:
                    item.quality = item.quality - item.quality

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
